### ISSUE TITLE
422 Unprocessable Entity While aws S3 as Storage Backend

### ISSUE THREAD

**Issue by alameenlr** [ankitpokhrel/tus-php#208]
**Setup Details**
 Storage : AWS S3 

**Errors**
PATCH _http://localhost/files/f586e634-1ca4-4d5b-b98c-1b46905f8f5f 422 (Unprocessable Entity)_
Uncaught Error: tus: _unexpected response while creating upload
    at XMLHttpRequest.t.onload_

**Comment by alameenlr** [comment]
You Don't Know Why it is Happened?

**Comment by ankitpokhrel** [comment]
@alameenlr From the error you posted I can only assume that the server is throwing some error. You need to check your server logs to figure out the actual issue.

**Comment by alameenlr** [comment]
PHP Fatal error:  Uncaught Error: Class 'TusPhp\Tus\Server' not found in /tusd/server.php:27
Stack trace:
#0 {main}  thrown in /tusd/server.php on line 27

**This is the Error Log**

**server.php**
`
require 'aws/autoload.php';
require 'tus/autoload.php';

use Aws\S3\S3Client;
use TusPhp\Tus\Server;
use Aws\Credentials\Credentials;

$awsAccessKey = 'AWS ACCESS KEY'; 
$awsSecretKey = 'AWS SECRET KEY';
$awsRegion    = 'us-east-2';
$basePath     = 's3://temp-bucket';
$endPoint = 'https://s3.us-east-2.amazonaws.com';

$s3Client = new S3Client(
	array(
		'endpoint' => $endPoint,
		'credentials' => array(
			'key' => $awsAccessKey,
			'secret' => $awsSecretKey 
		),
		'region' => $awsRegion,
		'version' => 'latest'
	)
);
$s3Client->registerStreamWrapper();

$server = new Server('file');
$server->setUploadDir($basePath);

$response = $server->serve();
$response->send();

exit(0);
`

**Comment by alameenlr** [comment]
![Capture](https://user-images.githubusercontent.com/50786913/69142081-1212c100-0aec-11ea-9321-c440a85532cd.JPG)


please Help Me, I can't found any error.
No error log found.

**Comment by samundra** [comment]
Hi @alameenlr,

From the above error post, it's very hard to debug what could have gone wrong. 

However, Above error could happen for following reasons:

1. File is not uploaded properly from client to tus-server
2. Server is rejecting uploaded file

We have to pin-point problem,  whether it's 1 or 2. Then, we can shift our focus to resolve it.

In order to that, We'll need following information from you.

- Please tell us about Tus client and Tus server setup. 
- Describe how you got above error? What action did you take?
- Share Javascript stack trace (You can expand PATCH request and uncaught error then share them)
- Request Headers for upload request
- Error Response from Server

Furthermore, it would also help if you can share server error logs. We can't give you exact steps as this solely depends on your server setup.

**Comment by navdbaloch** [comment]
Hi All,

I was also stuck with the same error.

After debugging, I found out that "FileException" was being thrown with the following message

"Unable to open {file path}"

Although the create request was a success, but the file was not created at the specified location.

Now looking into why the file was not created in the first request.

**Comment by alameenlr** [comment]
**tus Client** 
---------------------------------------
- _tus-js-client 1.8.0_
- _uploading function_
```js
	var _frm = $(this).parents('.form');
	var file = document.getelementbyid('files').files[0];
	console.log(file);
	var es_upload = {
		endpoint: '/store/files',
		retryDelays: [0, 1000, 3000, 5000],
		metadata: {
			name: file.name,
    			type: file.type
		},
		onProgress: function (bytesUploaded, bytesTotal){
			console.log(bytesUploaded,bytesTotal);
			_frm.find('.prgss').val((bytesUploaded*100)/bytesTotal);
		},
		onSuccess: function () {
			console.log('UPLOAD SUCCESS');
		}
	};
	var  tus_upload = new tus.Upload(file,es_upload);
	try{
        	tus_upload.start();
        }catch (e){
        	console.log('Tus error ' + e.name + ":" + e.message + "\n" + e.stack);
        }`
```
**tus Server**
---------------------------------------------------------------------
- _server.php_
```php
	<?php
		require 'tus/autoload.php';
		require 'aws/autoload.php';
		
		use Aws\S3\S3Client;
		use TusPhp\Tus\Server;
		use Aws\Credentials\Credentials;

		$awsAccessKey = 'AWS ACCESS KEY';
		$awsSecretKey = 'AWS SECRET KEY'; 

		$awsRegion    = 'us-east-2';
		$basePath     = 's3://temp-bucket';
		$endPoint = 'https://s3.us-east-2.amazonaws.com';

		$s3Client = new S3Client(
			array(
				'endpoint' => $endPoint,
				'credentials' => array(
					'key' => $awsAccessKey,
					'secret' => $awsSecretKey 
				),
				'region' => $awsRegion,
				'version' => 'latest'
			)
		);
		$s3Client->registerStreamWrapper();

		$server = new Server('file');
		$server->setUploadDir($basePath);

		$response = $server->serve();
		$response->send();

		exit(0);
	?>
```

- _http://localhost/store/files_
![Tus](https://user-images.githubusercontent.com/50786913/69216212-a4b96b80-0b91-11ea-8c07-7ae8cd7bb553.JPG)

- _Browser Console_
![tus2](https://user-images.githubusercontent.com/50786913/69216213-a5520200-0b91-11ea-9b0e-e8ecc18ac3a3.JPG)

- _Request Headers_
Accept: */*
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9
Cache-Control: no-cache
Connection: keep-alive
Content-Length: 663
Content-Type: application/offset+octet-stream
Cookie: x=JTdCJTIyYWNzX2
Host: localhost
Origin: http://localhost
Pragma: no-cache
Referer: http://localhost/
Sec-Fetch-Mode: cors
Sec-Fetch-Site: same-origin
Tus-Resumable: 1.0.0
Upload-Offset: 0
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.97 Safari/537.36

- _Response Headers_
Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Content-Length, Upload-Key, Upload-Checksum, Upload-Length, Upload-Offset, Tus-Version, Tus-Resumable, Upload-Metadata
Access-Control-Allow-Methods: GET,POST,PATCH,DELETE,HEAD,OPTIONS
Access-Control-Allow-Origin: http://localhost
Access-Control-Expose-Headers: Upload-Key, Upload-Checksum, Upload-Length, Upload-Offset, Upload-Metadata, Tus-Version, Tus-Resumable, Tus-Extension, Location
Access-Control-Max-Age: 86400
Cache-Control: no-cache, private
Connection: close
Content-Length: 48
Content-Type: text/html; charset=UTF-8
Date: Wed, 20 Nov 2019 06:46:12 GMT
Server: Apache/2.4.41 (Win64) OpenSSL/1.1.1c PHP/7.3.11
Tus-Resumable: 1.0.0
X-Content-Type-Options: nosniff
X-Powered-By: PHP/7.3.11

**Comment by ankitpokhrel** [comment]
@alameenlr Looks like you have not defined the route `/store/files` in your server. You need to define the route and upload the file from the route handler of that endpoint.

**Comment by samundra** [comment]
Hi @alameenlr 

First of all, thank you for your response. 

It seems like you have not defined route handler. The example that we provide in repository has this defined these routes internally. 

We have to define routes that will handle request to different paths. With Tus-php we'll have to define following routes.

```
# Gather information about server's current configuration
OPTIONS /files# Check if the given upload is valid
HEAD /files/{upload-key}# Create a new upload
POST /files# Resume upload created with POST
PATCH /files/{upload-key}# Delete the previous upload
DELETE /files/{upload-key}
```

Please verify that these routes are configured properly. Meaning that they are accessible and are functioning properly.

The request diagram looks as follows:

![Tus-php (3)](https://user-images.githubusercontent.com/760855/69224431-210e7780-0baf-11ea-9e9a-d8811a872ff1.jpg)

Here is article on PHP that describes about setting up Tus-php server. https://medium.com/hackernoon/resumable-file-upload-in-php-handle-large-file-uploads-in-an-elegant-way-e6c6dfdeaedb

Once we have finalised that routes are working properly and are returning valid response we can focus to debug any error that might have occurred on Tus-php server.

Please let us know, if you need further help with this.

**Comment by alameenlr** [comment]
What you mean by Routing.?
```
OPTIONS /files# Check if the given upload is valid
HEAD /files/{upload-key}# Create a new upload
POST /files# Resume upload created with POST
PATCH /files/{upload-key}# Delete the previous upload
DELETE /files/{upload-key}
```
All these now pointing to server.php

Cloud you an example for routing?,How did you identify route ```/store/files``` not defined?

**Comment by samundra** [comment]
@alameenlr 

you have defined endpoint in client side.

```
var es_upload = {
		endpoint: '/store/files',   <---------- This is the endpoint for tus-server
		retryDelays: [0, 1000, 3000, 5000],
		metadata: {
			name: file.name,
    			type: file.type
		},
		onProgress: function (bytesUploaded, bytesTotal){
			console.log(bytesUploaded,bytesTotal);
			_frm.find('.prgss').val((bytesUploaded*100)/bytesTotal);
		},
		onSuccess: function () {
			console.log('UPLOAD SUCCESS');
		}
	};
```

That endpoint `/store/file` should be accessible and return proper response.

**Comment by alameenlr** [comment]
That endpoint ```/store/file``` ?

```endpoint: '/store/files', // <--- This is the endpoint for Tus-server``` is it not right?
Could you please Send the Correct code for both server and client?

**Comment by samundra** [comment]
Hi @alameenlr 

Please refer to code example here for reference [Basic Example](https://github.com/ankitpokhrel/tus-php/tree/master/example/basic)

The setup of Tus-server and Tus-client requires a bit of knowledge on server setup as well. So, I would suggest you to get help from Sysadmin/Server person. These setup configurations is very very dynamic and depends on your server configuration so for that reason it's very hard for us to provide step by step guidelines for any particular setup. 

> endpoint: '/store/files', // <--- This is the endpoint for Tus-server is it not right?

I am confused what do you mean by is it not right. Can you please explain a bit?

Endpoint is simply a url which when accessed should be accessible and handle request as expected. In this case `/store/files` is expected to handle incoming upload request and send chunked response respectively.

**Comment by alameenlr** [comment]
Enough!!!!

**Comment by alameenlr** [comment]
Yah!. It is working on my local server.
Storing files on my local disk.
When I tried to put on AWS S3

![tus2](https://user-images.githubusercontent.com/50786913/69399170-2b498680-0d13-11ea-90b5-60f700ebdf38.JPG)

```php
	use Aws\S3\S3Client;
	use Aws\Credentials\Credentials;

	$awsAccessKey = ' AWS ACCESS KEY';
	$awsSecretKey = 'AWS SECRET KEY';

	$awsRegion    = 'us-east-2';
	$basePath     = 's3://temp-bucket';
	
	$s3Client = new S3Client(
		array(
			'region' => $awsRegion,
			'version' => 'latest',
			'credentials' => new Credentials($awsAccessKey, $awsSecretKey)
		)
	);
	$s3Client->registerStreamWrapper();

	$server = new TusPhp\Tus\Server('file');
	$server->setApiPath('/tusd');
	$server->setUploadDir($basePath);

	$response = $server->serve();
	$response->send();
```

**Comment by alameenlr** [comment]
Yah!. It is working on my local server.
Storing files on my local disk.
When I tried to put on AWS S3!

![tus2](https://user-images.githubusercontent.com/50786913/69432279-6b365b00-0d5f-11ea-8012-2dfbf0f5b937.JPG)

```php
//sever.php

	use Aws\S3\S3Client;
	use Aws\Credentials\Credentials;

	$awsAccessKey = ' AWS ACCESS KEY';
	$awsSecretKey = 'AWS SECRET KEY';

	$awsRegion    = 'us-east-2';
	$basePath     = 's3://temp-bucket';
	
	$s3Client = new S3Client(
		array(
			'region' => $awsRegion,
			'version' => 'latest',
			'credentials' => new Credentials($awsAccessKey, $awsSecretKey)
		)
	);
	$s3Client->registerStreamWrapper();

	$server = new TusPhp\Tus\Server('file');
	$server->setApiPath('/tusd');
	$server->setUploadDir($basePath);

	$response = $server->serve();
	$response->send();
```

**Comment by alameenlr** [comment]
https://github.com/ankitpokhrel/tus-php/commit/34bf866bdb067138d727b4141a3fa7e84c53580d 

Success!!!!

**Comment by alameenlr** [comment]
```
https://github.com/ankitpokhrel/tus-php/commit/34bf866bdb067138d727b4141a3fa7e84c53580d
```
Success

**Comment by ankitpokhrel** [comment]
Glad to know that you figured it out. Apologies that we didn't realize that commit was not yet released. It is included in the new release [v1.1.0](https://github.com/ankitpokhrel/tus-php/releases/tag/v1.1.0).
